### DISCUSSION TITLE
EMQX Message persistence for open source

### DISCUSSION THREAD

**Discussion by win5923** [emqx/emqx#12573] [Q&A]
Hello, I have some questions about Messenger persistence.

- using emqx open source , can we use external database as a storge for message persistence ? I've seen that RocksDB can be used for persistence, but how can it be configured in EMQX?
- in Kubernetes deployment using operator , how to guarantee that the messages will not be lost when restart the container ? I noticed that even with the volumeClaimTemplates configured, all messages are still cleared upon restarting EMQX.
- Can I persist retained messages through Data Bridge?

**Comment by thalesmg** [comment]
Hi!

> using emqx open source , can we use external database as a storge for message persistence ? I

Not really.  External data integrations are an Enterprise-only feature.  Community edition only offers HTTP and MQTT data bridges.

> I've seen that RocksDB can be used for persistence, but how can it be configured in EMQX?

Yes, that's a feature we're currently working on, it's currently experimental on alpha stage and might be subject to breaking changes between releases until it's stabilized.  If you are willing to help us test it, the basic configuration you may use to get started is (considering 5.5.0):

```hcl
session_persistence {
  enable = true
}
```

Note that the next release (5.6.0) will most likely change this configuration schema. 

> in Kubernetes deployment using operator , how to guarantee that the messages will not be lost when restart the container ? I noticed that even with the volumeClaimTemplates configured, all messages are still cleared upon restarting EMQX.

When the RocksDB persistence feature is stabilized, that and using persistent volumes in k8s will be the recommended way to ensure messages are stored between restarts.  

Note that, in the current versions, there's already a node evacuation feature that may be used in blue-green deployments and will migrate session data before the node is shut down.  For reference: https://docs.emqx.com/en/emqx-operator/latest/tasks/configure-emqx-blueGreenUpdate.html

> Can I persist retained messages through Data Bridge?

No, currently the only retainer backend is Mnesia.  You can, however, configure it to store messages to disc.  See https://docs.emqx.com/en/enterprise/latest/dashboard/retained.html#retainer-settings

**Comment by win5923** [comment]
Hi @thalesmg, so, only configuring `volumeClaimTemplates` alone won't retain messages. Need to set up RocksDB and `volumeClaimTemplates` together for that?

**Reply by thalesmg** [reply]
>  only configuring volumeClaimTemplates alone won't retain messages.

Correct.

> Need to set up RocksDB and volumeClaimTemplates together for that?

Yes.  

Note also that you can customize the RocksDB data directory with the following config (subject to change in future versions), so it's able to utilized a dedicated mounted volume if you want:

```hcl
session_persistence {
  storage {
    builtin {
      data_dir = "/data/dir"
    }
  }
}
```

**Reply by ennc0d3** [reply]
I tried with following configuration, in v5.6.1 (OpenSource) 

```
    durable_storage {
        messages {
           backend = "builtin"
        }
      }
```

When I check the the configuration via emqx ctl ds
```
emqx@emqx-core-99ddfbd86-0:/opt/emqx$ emqx ctl conf show durable_storage
durable_storage {
  messages {backend = builtin}
}

emqx@emqx-core-99ddfbd86-0:/opt/emqx$ emqx ctl ds
ds # Durable storage is disabled
```

Does this means durable_storage for message persistence is not supported?  I checked by publishing a message, restarted the core node ( Only 1 core in my case) and then subscribed to the topic, I didn't the message

**Reply by ieQu1** [reply]
Hi,

I replied in a different discussion. We've finished a documentation chapter that will hopefully answer all your questions. I'll link it to the documentation once it's online.

**Comment by ieQu1** [comment]
Hello everyone,

Apologies for withholding updates for a long time. EMQX v5.7, released today, contains first officially supported version of message persistence. Official name of the feature is "Durable Sessions".

Please check out the documentation here: https://www.emqx.io/docs/en/latest/durability/durability_introduction.html

This feature is included in both enterprise and opensource editions.
